agent/wsclient/client.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package wsclient wraps the generated aws-sdk-go client to provide marshalling
+// and unmarshalling of data over a websocket connection in the format expected
+// by backend. It allows for bidirectional communication and acts as both a
+// client-and-server in terms of requests, but only as a client in terms of
+// connecting.
+package wsclient
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4"
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/awslabs/aws-sdk-go/internal/protocol/json/jsonutil"
+	"github.com/gorilla/websocket"
+)
+
+var log = logger.ForModule("ws client")
```

Comment:
Not required, but it'd be nice to use this opportunity to properly migrate to seelog.